matrix-sdk/src/main/java/org/matrix/androidsdk/login/AutoDiscovery.kt

```diff
@@ -0,0 +1,191 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.matrix.androidsdk.login
+
+import android.net.Uri
+import com.google.gson.stream.MalformedJsonException
+import org.json.JSONObject
+import org.matrix.androidsdk.HomeServerConnectionConfig
+import org.matrix.androidsdk.rest.callback.ApiCallback
+import org.matrix.androidsdk.rest.client.AutoDiscoveryRestClient
+import org.matrix.androidsdk.rest.client.IdentityPingRestClient
+import org.matrix.androidsdk.rest.client.LoginRestClient
+import org.matrix.androidsdk.rest.model.HttpException
+import org.matrix.androidsdk.rest.model.MatrixError
+import org.matrix.androidsdk.rest.model.Versions
+import org.matrix.androidsdk.rest.model.WellKnown
+import java.io.EOFException
+import java.lang.Exception
+import java.net.MalformedURLException
+import java.net.URL
+import java.net.UnknownHostException
+import javax.net.ssl.HttpsURLConnection
+
+
+class AutoDiscovery {
+
+    private var discoveryRestClient = AutoDiscoveryRestClient()
+
+    data class DiscoveredClientConfig(
+            val action: Action,
+            val wellKnown: WellKnown? = null
+    )
+
+    enum class Action {
+        /*
+         Retrieve the specific piece of information from the user in a way which fits within the existing client user experience,
+         if the client is inclined to do so. Failure can take place instead if no good user experience for this is possible at this point.
+         */
+        PROMPT,
+        /*
+         Stop the current auto-discovery mechanism. If no more auto-discovery mechanisms are available,
+         then the client may use other methods of determining the required parameters, such as prompting the user, or using default values.
+         */
+        IGNORE,
+        /*
+            Inform the user that auto-discovery failed due to invalid/empty data and PROMPT for the parameter.
+         */
+        FAIL_PROMPT,
+        /*
+         Inform the user that auto-discovery did not return any usable URLs. Do not continue further with the current login process.
+         At this point, valid data was obtained, but no homeserver is available to serve the client.
+         No further guess should be attempted and the user should make a conscientious decision what to do next.
+         */
+        FAIL_ERROR,
+
+    }
+
+
+    fun findClientConfig(domain: String, callback: ApiCallback<DiscoveredClientConfig>) {
+
+        discoveryRestClient.getWellKnown(domain, object : ApiCallback<WellKnown> {
```

Comment:
You can use `SimpleApiCallback(callback)` here to avoid overriding `onUnexpectedError` and `onMatrixError`

Reply:
Wanted to but it's showing Toast or i have to provide failure callback which requires to implement methods anyway